giraffe/src/utils/orderBy.ts

```diff
@@ -0,0 +1,188 @@
+  const length = objCriteria.length
+  const ordersLength = orders.length
+
+  while (++index < length) {
+    const order = index < ordersLength ? orders[index] : null
+    const cmpFn =
+      order && typeof order === 'function' ? order : compareAscending
+    const result = cmpFn(objCriteria[index], othCriteria[index])
+    if (result) {
+      if (order && typeof order !== 'function') {
+        return result * (order == 'desc' ? -1 : 1)
+      }
+      return result
+    }
+  }
+  return object.index - other.index
+}
+
+const isArrayLike = value => {
+  return value != null && typeof value !== 'function' && isLength(value.length)
+}
+
+const baseFor = (object, iteratee, keysFunc) => {
+  const iterable = Object(object)
+  const props = keysFunc(object)
+  let {length} = props
+  let index = -1
+
+  while (length--) {
+    const key = props[++index]
+    if (iteratee(iterable[key], key, iterable) === false) {
+      break
+    }
+  }
+  return object
+}
+
+const baseForOwn = (object, iteratee) => {
+  return object && baseFor(object, iteratee, Object.keys)
+}
+
+const baseEach = (collection, iteratee) => {
+  if (collection == null) {
+    return collection
+  }
+  if (!isArrayLike(collection)) {
+    return baseForOwn(collection, iteratee)
+  }
+  const length = collection.length
+  const iterable = Object(collection)
+  let index = -1
+
+  while (++index < length) {
+    if (iteratee(iterable[index], index, iterable) === false) {
+      break
+    }
+  }
+  return collection
+}
+
+const baseOrderBy = (collection, iteratees, orders) => {
+  if (iteratees.length) {
+    iteratees = iteratees.map(iteratee => {
+      if (Array.isArray(iteratee)) {
+        return value =>
+          get(value, iteratee.length === 1 ? iteratee[0] : iteratee)
+      }
+
+      return iteratee
+    })
+  } else {
+    iteratees = [identity]
+  }
+
+  let criteriaIndex = -1
+  let eachIndex = -1
+
+  const result = Array.isArray(collection) ? new Array(collection.length) : []
+
+  baseEach(collection, value => {
+    const criteria = iteratees.map(iteratee => {
+      if (isString(iteratee)) {
+        return typeof value === 'function' ? value(iteratee) : value[iteratee]
+      }
+      return iteratee(value)
+    })
+
+    result[++eachIndex] = {
+      criteria,
+      index: ++criteriaIndex,
+      value,
+    }
+  })
+
+  return baseSortBy(result, (object, other) =>
+    compareMultiple(object, other, orders)
+  )
+}
+
+type OrderByIteratees =
+  | Function
+  | Array<string>
+  | Array<Function>
+  | Array<Array<string>>
+  | Array<object>
+
+export const orderBy = <X>(
+  collection: Array<X> | object,
+  iteratees: OrderByIteratees,
+  orders: Array<string>
+): Array<X> => {
+  if (collection == null) {
+    return []
+  }
+  if (!Array.isArray(iteratees)) {
+    iteratees = iteratees == null ? [] : [iteratees]
+  }
+  if (!Array.isArray(orders)) {
+    orders = orders == null ? [] : [orders]
+  }
+  return baseOrderBy(collection, iteratees, orders)
+}
```

Comment:
i'm a little bit worried that this might be a performance bottleneck - there's quite a bit of looping within looping happening here, i think. i think it might be a good idea for you and me and boatwright to sit down and look at the perf implications of this.

Reply:
This comes from lodash, which we already use in the UI's implementation of tables. So, if there is a performance bottleneck, it already exists. I think the goal here is just to move tables into Giraffe. If we want to drill for every ounce of performance, we should do that separately in another story. Or maybe an epic that encompasses performance everywhere in Giraffe.